### PR TITLE
HTTP server websocket support

### PR DESCRIPTION
**Pull request by jukkar**
Changes by this PR:
* Upgrade the HTTP server connection to a websocket connection
* Create support for websocket server
* Add websocket server support to the sample application

### CODE REVIEW

**Comment by jukkar** [comment]
This script can be used to connect to the sample http-server and upgrade the connection
```
#!/usr/bin/env python3

import websocket

websocket.enableTrace(True)
ws = websocket.WebSocket()
ws.connect("ws://192.0.2.1/")
ws.send("Hello, Server")
print(ws.recv())
while True:
    line = input("> ")
    if line == "quit":
        break
    ws.send(line)
    print(ws.recv())

ws.close()
```

**Review by cfriedt — DISMISSED** [review]
@jukkar - wow - that was fast!

File: subsys/net/lib/http/http_server_http1.c [hunk 1]
```diff
@@ -443,7 +468,34 @@ int handle_http1_request(struct http_client_ctx *client)
...
+		if (client->websocket_upgrade) {
+			goto upgrade_not_found;
+		}
+
+		if (client->http2_upgrade) {
+			return handle_http1_to_http2_upgrade(client);
+		}
+
+upgrade_not_found:
+		(void)zsock_send(client->fd, upgrade_required,
```
Comment (rlubos) [inline subsys/net/lib/http/http_server_http1.c:507]: Shouldn't we stick to `http_server_sendall()`? In theory, `zsock_send()` may not send al the requested data.

File: subsys/net/lib/http/http_server_http1.c [hunk 2]
```diff
@@ -314,11 +314,35 @@ static int on_header_field(struct http_parser *parser, const char *at,
 						   struct http_client_ctx,
 						   parser);
 
+	if (strncmp(at, "Upgrade", 7) == 0) {
```
Comment (rlubos) [inline subsys/net/lib/http/http_server_http1.c:317]: Such an approach is not 100% robust, HTTP1 parser's `http_data_cb()` does not guarantee to provide complete header field/value in a single call. I've experienced this when testing with very small RX buffer sizes, in case an incomplete header is read with `zsock_recv()`, the parser will call the callback twice (or even more for large fields). This case has been addressed in `on_url()` cb, we should do the same for headers (prob need some transient buffer).

File: subsys/net/lib/http/http_server_http1.c [hunk 3]
```diff
@@ -443,7 +468,34 @@ int handle_http1_request(struct http_client_ctx *client)
...
 
 	if (client->has_upgrade_header) {
-		return handle_http1_to_http2_upgrade(client);
+		static const char upgrade_required[] =
+			"HTTP/1.1 426 Upgrade required\r\n"
+			"Upgrade: ";
+		static const char upgrade_msg[] =
+			"Content-Length: 13\r\n\r\n"
+			"Wrong upgrade";
+		char *needed_upgrade = "h2c\r\n";
```
Comment (rlubos) [inline subsys/net/lib/http/http_server_http1.c:489]: `const char *`

File: subsys/net/lib/http/http_server_ws.c [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	ret = base64_encode(tmp, sizeof(tmp) - 1, &olen, accept, sizeof(accept));
+	if (ret) {
+		if (ret == -ENOMEM) {
+			NET_DBG("[%p] Too short buffer olen %zd", client, olen);
+		}
+
+		goto error;
+	}
+
+	(void)zsock_send(client->fd, switching_protocols,
```
Comment (rlubos) [inline subsys/net/lib/http/http_server_ws.c:66]: Shall we use `http_server_sendall()` here as well?

File: subsys/net/lib/http/http_server_ws.c [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+/* Handle upgrade from HTTP/1.1 to Websocket, see RFC 6455
+ */
+int handle_http1_to_websocket_upgrade(struct http_client_ctx *client)
+{
+	static const char switching_protocols[] =
+		"HTTP/1.1 101 Switching Protocols\r\n"
+		"Connection: Upgrade\r\n"
+		"Upgrade: websocket\r\n"
+		"Sec-WebSocket-Accept: ";
+	char key_accept[32 + sizeof(WS_MAGIC)];
```
Comment (rlubos) [inline subsys/net/lib/http/http_server_ws.c:41]: Is this 32 related to `HTTP_SERVER_WS_MAX_SEC_KEY_LEN`?

File: subsys/net/lib/http/http_server_http1.c [hunk 6]
```diff
@@ -489,7 +489,10 @@ int handle_http1_request(struct http_client_ctx *client)
 		char *needed_upgrade = "h2c\r\n";
 
 		if (client->websocket_upgrade) {
-			if (IS_ENABLED(CONFIG_WEBSOCKET)) {
+			if (IS_ENABLED(CONFIG_HTTP_SERVER_WEBSOCKET)) {
+				client->current_detail =
+					get_resource_detail(client->url_buffer,
+							    &path_len, true);
```
Comment (rlubos) [inline subsys/net/lib/http/http_server_http1.c:495]: We don't seem to handle well the situation when no websocket resource has been found. No check is here, and `handle_http1_to_websocket_upgrade()` seems to dereference `client->current_detail` unconditionally.

Personally, I'd also reconsider whether there's a need to modify `get_resource_detail()`.  `handle_http1_to_websocket_upgrade()` could simply check the resource type and report an error if the resource is of a different type. That way we keep the websocket processing details withing the websocket part of the server implementation.

**Review by rlubos — COMMENTED** [review]
Nice work!

File: subsys/net/lib/http/http_server_http1.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_http1.c:507]: I was using `zsock_send()` here intentionally so that we could avoid sending partial message, there is `http_server_sendall()` few lines below that will make sure the headers are then fully sent.

File: subsys/net/lib/http/http_server_http1.c (hunk 2 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_http1.c:317]: Indeed, good point, will fix.

File: subsys/net/lib/http/http_server_http1.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_http1.c:507]: Hmm, I just realized that I was wrong here, I need to change this to use sendall version.

File: subsys/net/lib/http/http_server_ws.c (hunk 5 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_ws.c:41]: Yes, that is correct. I will change this to use the macro.

File: samples/net/sockets/http_server/Kconfig [hunk 7]
```diff
@@ -38,4 +38,16 @@ config NET_SAMPLE_CERTS_WITH_SC
 	  Enable this flag, if you are interested to run this
 	  application with signed certificates and keys.
 
+config NET_SAMPLE_WEBSOCKET_SERVICE
```
Comment (kartben) [inline samples/net/sockets/http_server/Kconfig:41]: please remember to also update the README to mention WebSocket: at a minimum, this new option should be mentioned in "Server Customization" section, but also would be nice to have some basic instructions on how to "play" with the websocket.

BTW, but I know that you have it on your list already to polish the documentation, the README should likely also indicate the various endpoints and associated methods that are exposed by the server, e.g. `/dynamic` would be worth mentioning/explaining.

**Review by kartben — CHANGES_REQUESTED** [review]

**Comment by jukkar** [comment]
* Updated according to comments

**Review by kartben — DISMISSED** [review]
+1 for the sample's doc, thanks!

**Comment by jukkar** [comment]
* CI fixes

**Review by rlubos — DISMISSED** [review]
LGTM

**Review by kartben — DISMISSED** [review]

File: subsys/net/lib/http/http_server_core.c [hunk 8]
```diff
@@ -625,11 +632,35 @@ static int compare_strings(const char *s1, const char *s2)
...
+		if (detail->type != HTTP_RESOURCE_TYPE_WEBSOCKET) {
+			return true;
+		}
+	} else {
+		if (detail->type == HTTP_RESOURCE_TYPE_WEBSOCKET) {
+			return true;
+		}
+	}
+
+	return false;
```
Comment (pdgendt) [inline subsys/net/lib/http/http_server_core.c:651]: Maybe less readable, but can be a oneline.

```c
	return is_websocket != (detail->type == HTTP_RESOURCE_TYPE_WEBSOCKET);
```

File: subsys/net/lib/http/http_server_http1.c [hunk 9]
```diff
@@ -313,12 +313,86 @@ static int on_header_field(struct http_parser *parser, const char *at,
 	struct http_client_ctx *ctx = CONTAINER_OF(parser,
 						   struct http_client_ctx,
 						   parser);
+	size_t offset = strlen(ctx->header_buffer);
```
Comment (pdgendt) [inline subsys/net/lib/http/http_server_http1.c:316]: Should never occur, but I always tend to prefer the `n` variant when using fixed sized buffers. (If for some obscure reason the trailing `\0` would have been removed).
```suggestion
	size_t offset = strnlen(ctx->header_buffer, sizeof(ctx->header_buffer));
```

File: subsys/net/lib/http/http_server_http1.c [hunk 10]
```diff
@@ -313,12 +313,86 @@ static int on_header_field(struct http_parser *parser, const char *at,
...
+				}
+
+				ctx->has_upgrade_header = false;
+			}
+
+			if (ctx->websocket_sec_key_next) {
+#if defined(CONFIG_WEBSOCKET)
+				strncpy(ctx->ws_sec_key, ctx->header_buffer,
+					MIN(sizeof(ctx->ws_sec_key),
+					    strlen(ctx->header_buffer)));
```
Comment (pdgendt) [inline subsys/net/lib/http/http_server_http1.c:384]: The `MIN` macro would expand the `strlen` twice, can't use `offset` instead?

```suggestion
				strncpy(ctx->ws_sec_key, ctx->header_buffer,
					MIN(sizeof(ctx->ws_sec_key), offset));
```

**Review by pdgendt — DISMISSED** [review]
Just some nits

File: subsys/net/lib/http/http_server_core.c (hunk 8 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_core.c:651]: I would leave this as is, your suggestion is shorter but my brain just halts when trying to decipher the what is going on here.

File: subsys/net/lib/http/http_server_http1.c (hunk 9 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_http1.c:316]: The buffer should be always null terminated, and even if there null is missing, we have a check in next line to verify that the offset is not overflowing. In that case we also get a printout that something is wrong which we would not get with `strnlen`. I would suggest to leave this as is.

File: subsys/net/lib/http/http_server_http1.c (hunk 10 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_http1.c:384]: Good point, we could use offset here.

File: subsys/net/lib/http/http_server_http1.c (hunk 9 above)
Comment (pdgendt) [reply subsys/net/lib/http/http_server_http1.c:316]: The focus is on "should be always null terminated" even reading past the buffer end (which `strlen` will do) can lead to undefined behavior. For example with some `MPU` configurations.

I'm not saying there is an error here, it's just defensive programming.
Oversights might happen if someone were to add code between the check and the handling.

Comment (pdgendt) [reply subsys/net/lib/http/http_server_http1.c:316]: There's a reason why modern languages keep length information tied to arrays/slices.

**Review by rlubos — DISMISSED** [review]

File: subsys/net/lib/http/http_server_http1.c (hunk 9 above)
Comment (jukkar) [reply subsys/net/lib/http/http_server_http1.c:316]: Ok, if you feel we need this check, then it can be added of course.

**Review by pdgendt — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
